File: Python/10-BitManipulation/Problems/3-ReverseBits/ReverseBits.py
def reverse_bits_method1(n: int) -> int:
    """
    Method 1: Iterative approach using bitwise operations.

    This method iterates through the bits of the input integer,
    extracting each bit and placing it in the corresponding reversed
    position in the result.

    Args:
        n: The integer whose bits are to be reversed.

    Returns:
        The integer with the bits reversed.
    """
    reversed_n = 0
    for i in range(32):  # Assuming 32-bit integer
        # Get the rightmost bit of n
        bit = (n >> i) & 1
        # Place the bit in the correct position in reversed_n
        reversed_n = reversed_n | (bit << (31 - i))
    return reversed_n

def reverse_bits_method2(n: int) -> int:
    """
    Method 2: Using string manipulation.

    This method converts the integer to its binary string representation,
    reverses the string, and converts it back to an integer.

    Args:
        n: The integer whose bits are to be reversed.

    Returns:
        The integer with the bits reversed.
    """
    binary_string = bin(n)[2:].zfill(32)  # Remove '0b' prefix and pad with zeros
    reversed_string = binary_string[::-1]  # Reverse the string
    return int(reversed_string, 2)  # Convert the reversed string to an integer

def reverse_bits_method3(n: int) -> int:
    """
    Method 3: Divide and conquer approach.

    This method uses a divide and conquer strategy to reverse the bits.
    It swaps pairs, then groups of 4, then 8, 16.

    Args:
        n: The integer whose bits are to be reversed.

    Returns:
        The integer with the bits reversed.
    """
    n = (n >> 16) | (n << 16)
    n = ((n & 0xFF00FF00) >> 8) | ((n & 0x00FF00FF) << 8)
    n = ((n & 0xF0F0F0F0) >> 4) | ((n & 0x0F0F0F0F) << 4)
    n = ((n & 0xCCCCCCCC) >> 2) | ((n & 0x33333333) << 2)
    n = ((n & 0xAAAAAAAA) >> 1) | ((n & 0x55555555) << 1)
    return n

def reverse_bits_method4(n: int) -> int:
    """
    Method 4: Using bit manipulation with a loop and pop.

    This method uses a while loop to process the bits of the integer
    and constructs the reversed integer bit by bit.  It's similar to method 1,
    but perhaps more clear to understand for some.

    Args:
        n: The integer to reverse.

    Returns:
        The reversed integer.
    """
    reversed_num = 0
    count = 32
    while count > 0:
        reversed_num <<= 1  # left shift
        # Get the rightmost bit
        reversed_num |= (n & 1)
        n >>= 1  # Right shift
        count -= 1
    return reversed_num

def reverse_bits_method5(n: int) -> int:
    """
    Method 5: Using built-in bit_length and a loop.

    This method calculates the number of significant bits, and then iterates
    through that many bits, constructing the reversed number.  It handles
    numbers with less than 32 significant bits correctly.

    Args:
        n: The integer to reverse

    Returns:
        The reversed integer.
    """
    length = n.bit_length()
    reversed_num = 0
    for i in range(length):
        if (n >> i) & 1:
            reversed_num |= 1 << (length - 1 - i)
    return reversed_num


if __name__ == "__main__":
    test_number = 43261596  # Example input
    print(f"Original Number: {test_number}")

    # Method 1
    reversed_number1 = reverse_bits_method1(test_number)
    print(f"Reversed (Method 1): {reversed_number1}")  # Output: 964176192

    # Method 2
    reversed_number2 = reverse_bits_method2(test_number)
    print(f"Reversed (Method 2): {reversed_number2}")  # Output: 964176192

    # Method 3
    reversed_number3 = reverse_bits_method3(test_number)
    print(f"Reversed (Method 3): {reversed_number3}")  # Output: 964176192

    # Method 4
    reversed_number4 = reverse_bits_method4(test_number)
    print(f"Reversed (Method 4): {reversed_number4}")  # Output: 964176192

    # Method 5
    reversed_number5 = reverse_bits_method5(test_number)
    print(f"Reversed (Method 5): {reversed_number5}")  # Output: 964176192

    test_number2 = 0b00000010100101000001111010011100
    print(f"\nOriginal Number: {test_number2}")
    reversed_number5_2 = reverse_bits_method5(test_number2)
    print(f"Reversed (Method 5) binary: {reversed_number5_2}")

    test_number3 = 1  # Example with only one bit set
    print(f"\nOriginal Number: {test_number3}")
    reversed_number5_3 = reverse_bits_method5(test_number3)
    print(f"Reversed (Method 5) one bit: {reversed_number5_3}") # Output: 2147483648

File: Java/10-BitManipulation/Problems/3-ReverseBits/ReverseBits.java
public class ReverseBits {

    /**
     * **Approach 1: Iterative Bitwise Swap**
     *
     * This method iterates through the 32 bits of the integer. In each iteration,
     * it swaps the i-th bit from the left with the (31-i)-th bit from the right.
     * It uses bitwise AND to check if a bit is set, and bitwise OR to set a bit
     * in the result.  It's generally efficient as it only iterates 16 times.
     *
     * @param n The integer whose bits are to be reversed
     * @return The integer with reversed bits
     */
    public int reverseBits1(int n) {
        int result = 0;
        for (int i = 0; i < 16; i++) {
            // Get the i-th bit from the right
            int rightBit = (n >> i) & 1;
            // Get the i-th bit from the left
            int leftBit = (n >> (31 - i)) & 1;

            // Set the i-th bit from the right in the result
            result |= (leftBit << i);
            // Set the i-th bit from the left in the result
            result |= (rightBit << (31 - i));
        }
        return result;
    }

    /**
     * **Approach 2: Bitwise Operations and Shifting**
     *
     * This approach is optimized for speed.  It processes the bits one by one,
     * taking the least significant bit of the input, and placing it in the
     * most significant bit of the result.  It shifts the result to the left
     * and the input to the right in each step.
     *
     * @param n The integer to reverse
     * @return The reversed integer
     */
    public int reverseBits2(int n) {
        int result = 0;
        for (int i = 0; i < 32; i++) {
            // Get the rightmost bit of n
            int rightBit = n & 1;
            // Shift the current result to the left, making space for the new bit.
            result <<= 1;
            // Place the rightmost bit of n into the least significant bit of result
            result |= rightBit;
            // Discard the rightmost bit of n
            n >>>= 1; // Use unsigned right shift to avoid sign extension
        }
        return result;
    }

    /**
     * **Approach 3: Using Integer.reverse()**
     *
     * This is the simplest and most efficient approach in Java.  The
     * `Integer.reverse()` method is a built-in function that does exactly
     * what we need.  It's highly optimized at the JVM level.
     *
     * @param n The integer to reverse
     * @return The reversed integer
     */
    public int reverseBits3(int n) {
        return Integer.reverse(n);
    }

    /**
     * **Approach 4:  Lookup Table (for repeated calls)**
     *
     * If the reverseBits function is called many times, a lookup table can
     * improve performance.  This approach pre-computes the reversed values
     * for all possible byte values (0-255) and stores them in an array.  Then,
     * the input integer is processed byte by byte, and the reversed value for
     * each byte is retrieved from the table.  This trades space for time.
     *
     * @param n The integer to reverse
     * @return The reversed integer
     */
    private static final int[] reverseTable = new int[256];

    // Static initializer to pre-compute the reversed bits for each byte
    static {
        for (int i = 0; i < 256; i++) {
            reverseTable[i] = reverseByte(i);
        }
    }

    private static int reverseByte(int b) {
        int result = 0;
        for (int i = 0; i < 8; i++) {
            result |= ((b >> i) & 1) << (7 - i);
        }
        return result;
    }

    public int reverseBits4(int n) {
        int result = 0;
        // Process the integer in 4 bytes.
        result |= reverseTable[(n & 0xFF)] << 24;
        result |= reverseTable[((n >> 8) & 0xFF)] << 16;
        result |= reverseTable[((n >> 16) & 0xFF)] << 8;
        result |= reverseTable[((n >> 24) & 0xFF)];
        return result;
    }

    /**
     * **Approach 5: Divide and Conquer**
     *
     * This approach uses a divide-and-conquer strategy to reverse the bits. It performs a series of bitwise swaps
     * to reverse the bits in pairs, then groups of 4, then groups of 8, 16
     * @param n the integer to be reversed
     * @return the reversed integer
     */
    public int reverseBits5(int n) {
        // Swap adjacent bits
        n = (n & 0x55555555) << 1 | (n >>> 1) & 0x55555555;
        // Swap pairs of bits
        n = (n & 0x33333333) << 2 | (n >>> 2) & 0x33333333;
        // Swap groups of 4 bits
        n = (n & 0x0F0F0F0F) << 4 | (n >>> 4) & 0x0F0F0F0F;
        // Swap groups of 8 bits
        n = (n & 0x00FF00FF) << 8 | (n >>> 8) & 0x00FF00FF;
        // Swap groups of 16 bits
        n = (n << 16) | (n >>> 16);
        return n;
    }

    public static void main(String[] args) {
        ReverseBits solution = new ReverseBits();
        int num = 43261596; // Example input

        System.out.println("Original Number: " + num);
        System.out.println("Reversed Bits (Approach 1): " + solution.reverseBits1(num));
        System.out.println("Reversed Bits (Approach 2): " + solution.reverseBits2(num));
        System.out.println("Reversed Bits (Approach 3): " + solution.reverseBits3(num));
        System.out.println("Reversed Bits (Approach 4): " + solution.reverseBits4(num));
        System.out.println("Reversed Bits (Approach 5): " + solution.reverseBits5(num));
    }
}
